### ISSUE TITLE
delegate_to and (haproxy) module, incorrect results

### ISSUE THREAD

**Issue by hyperized** [ansible/ansible#14741]
##### Issue Type:
- Bug Report
##### Ansible Version:

```
ansible --version
ansible 2.0.1.0
  config file = /home/user/ansible.cfg
  configured module search path = /usr/share/ansible
```
##### Ansible Configuration:

N/A
##### Environment:

N/A
##### Summary:

I seem to have an issue with the delegate_to and the use of modules (haproxy in this case). The behavior I observe is that the delegated task is only executed once, not the expected two or more times.

The issue at hand is that this solution worked for Ansible 1.9, 2.0 up until 2.0.1.0.
I'm fairly sure this problem is _not_ in the scope of the haproxy module, but I have no other good example to verify that with, but I cannot exclude it 100% either.

So the following playbook is used to drain the backend in haproxy, should target each webserver and should on behalf of the webserver remove itself from the haproxy loadbalancers (pretty straight foward). Using 'run_once' makes no difference in the actual result.
##### Steps To Reproduce:

```

---
- hosts: webservers #
  serial: 1
  gather_facts: true
  max_fail_percentage: 0
  become: yes
  vars:
    haproxy_backend: myspecialbackend
  pre_tasks:
    - name: drain in haproxy
      haproxy:
        backend: "{{ haproxy_backend }}"
        host: "{{ ansible_hostname }}" # Short hostname of webserver
        shutdown_sessions: no
        socket: "{{ haproxy_socket_location }}"
        state: disabled
        wait: yes
        wait_interval: "{{ haproxy_mode_change_wait }}"
        wait_retries: 25
      register: results
      delegate_to: "{{ item }}"
      delegate_facts: false # Ensure this 2.0 feature is not used
      run_once: true
      with_items: "{{ groups.tag_group_haproxy }}" # Valid selection of EC2 haproxy instances

    - name: Show results
      debug:
        msg: "{{ results }}"
```

I redacted the output to not show any system specific information such as IP addresses.
The result is pretty much what you'd expect:

```
TASK [setup] *******************************************************************
ok: [web01]

TASK [drain in haproxy] ********************************************************
changed: [web01 -> haproxy01] => (item=haproxy01)
ok: [web01 -> haproxy02] => (item=haproxy02)

TASK [Show results] ************************************************************
ok: [web01] => {
    "msg": {
        "changed": true,
        "msg": "All items completed",
        "results": [
            {
                "_ansible_delegated_vars": {
                    "ansible_host": "haproxy01"
                },
                "_ansible_no_log": false,
                "changed": true,
                "invocation": {
                    "module_args": {
                        "backend": "myspecialbackend",
                        "host": "web01",
                        "shutdown_sessions": false,
                        "socket": "/var/run/haproxy.sock",
                        "state": "disabled",
                        "wait": true,
                        "wait_interval": 15,
                        "wait_retries": 25,
                        "weight": null
                    },
                    "module_name": "haproxy"
                },
                "item": "haproxy01",
                "stdout": "...[REDACTED]
                myspecialbackend,web01,0,0,0,1,,8,20344,112614,,0,,0,0,0,0,MAINT,1,1,0,1,2,0,74,,1,5,1,,0,,2,0,,1,L7OK,200,4,0,8,0,0,0,0,0,,,,0,0,,,,,424,,,0,1,5,5,
                ...",
                "stdout_lines": [
                    "...[REDACTED]
                    "myspecialbackend,web01,0,0,0,1,,8,20344,112614,,0,,0,0,0,0,MAINT,1,1,0,1,2,0,74,,1,5,1,,0,,2,0,,1,L7OK,200,4,0,8,0,0,0,0,0,,,,0,0,,,,,424,,,0,1,5,5,",
                ]
            },
            {
                "_ansible_delegated_vars": {
                    "ansible_host": "haproxy02"
                },
                "_ansible_no_log": false,
                "changed": false,
                "invocation": {
                    "module_args": {
                        "backend": "myspecialbackend",
                        "host": "web01",
                        "shutdown_sessions": false,
                        "socket": "/var/run/haproxy.sock",
                        "state": "disabled",
                        "wait": true,
                        "wait_interval": 15,
                        "wait_retries": 25,
                        "weight": null
                    },
                    "module_name": "haproxy"
                },
                "item": "haproxy02",
                "stdout": "...[REDACTED]
                myspecialbackend,web01,0,0,0,1,,8,20344,112614,,0,,0,0,0,0,MAINT,1,1,0,1,2,0,74,,1,5,1,,0,,2,0,,1,L7OK,200,4,0,8,0,0,0,0,0,,,,0,0,,,,,424,,,0,1,5,5,
                ...",
                "stdout_lines": [
                    "myspecialbackend,web01,0,0,0,1,,8,20344,112614,,0,,0,0,0,0,MAINT,1,1,0,1,2,0,74,,1,5,1,,0,,2,0,,1,L7OK,200,4,0,8,0,0,0,0,0,,,,0,0,,,,,424,,,0,1,5,5,",
                ]
            }
        ]
    }
}

PLAY RECAP *********************************************************************
web01               : ok=3    changed=1    unreachable=0    failed=0
```
##### Expected Results:

I expected both haproxys to obtain a MAINT state for the both backends.
##### Actual Results:

One of the haproxy instances has a changed state, the other does not.

Now if you check on the actual haproxy instance, no change is executed at all. It seems that the module is reporting the changes made on haproxy01 instead of haproxy02 (note the identical counter numbers, thats extremely unlikely). 

I can theorize that this problem is either that the delegate somehow saves the old results and returns these to the module or the haproxy module is faulty, I cannot exclude which from my position.

**Comment by hyperized** [comment]
Thanks for delegating the bug ;)

**Comment by rfleschenberg** [comment]
I am seeing a similar problem, without the haproxy module. It seems that currently, `delegate_to` in conjunction with `with_items` is broken.

Here is a minimal example https://github.com/rfleschenberg/ansible-delegation-bug

**Comment by rfleschenberg** [comment]
Note: I am not using ansible_ssh_hostname there.

**Comment by rfleschenberg** [comment]
I am not familiar with Ansible's codebase, but I poked around with a debugger a little. I think the problem is here, or in some code more or less closely related to this:

https://github.com/ansible/ansible/blob/devel/lib/ansible/playbook/play_context.py#L365-L367

The `variables` mapping passed to the `set_task_and_variable_override()` method contains the keys `ansible_host` and `ansible_ssh_host`, both mapping to the wrong host name (in my case, the one from the earlier loop iteration). This code then sets the wrong host name as `new_info.remote_addr`. The correct value is present in `delegated_vars` and seems to be ignored. I am not sure what the correct fix is.

**Comment by rfleschenberg** [comment]
I think the problem is related to the `ansible_host` / `ansible_ssh_host` ambiguity. Only the new name `ansible_host` is present in `delegated_vars`, so the check on line 362 fails for `ansible_ssh_host` and the code falls through to line 365. Reversing the order of the tuple on line 53 so that `ansible_host` is checked first works around the problem, but I guess this applies to other deprecated variables as well and you probably want a cleaner fix :)

**Comment by bcoca** [comment]
please confirm that #15024 resolves this issue for you

**Comment by rfleschenberg** [comment]
Yep, #15024 works for me. Thanks!

**Comment by bcoca** [comment]
closing via merged #15024
